core/templates/pages/exploration-editor-page/services/exploration-states.service.ts

```diff
@@ -18,571 +18,759 @@
+        '\nAll states names: ' + this._states.getStateNames());
+      e.message += additionalInfo;
+      throw e;
+    }
 
-    var _setState = function(stateName, stateData, refreshGraph) {
-      _states.setState(stateName, angular.copy(stateData));
-      if (refreshGraph) {
-        refreshGraphEventEmitter.emit();
-      }
-    };
-
-    var getStatePropertyMemento = function(stateName, backendName) {
-      var accessorList = PROPERTY_REF_DATA[backendName];
-      var propertyRef = _states.getState(stateName);
-      try {
-        accessorList.forEach(function(key) {
-          propertyRef = propertyRef[key];
+    return cloneDeep(propertyRef);
+  }
+
+  saveStateProperty(
+      stateName: string, backendName: 'content', newValue: SubtitledHtml
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'param_changes', newValue: ParamChange[]
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'widget_id', newValue: string
+  ): void;
+  saveStateProperty(
+      stateName: string,
+      backendName: 'widget_customization_args',
+      newValue: InteractionCustomizationArgs
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'answer_groups', newValue: AnswerGroup[]
+  ): void;
+  saveStateProperty(
+      stateName: string,
+      backendName: 'confirmed_unclassified_answers',
+      newValue: AnswerGroup[]
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'default_outcome', newValue: Outcome
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'hints', newValue: Hint[]
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'solution', newValue: SubtitledHtml
+  ): void;
+  saveStateProperty(
+      stateName: string,
+      backendName: 'recorded_voiceovers',
+      newValue: RecordedVoiceovers
+  ): void;
+  saveStateProperty(
+      stateName: string,
+      backendName: 'solicit_answer_details',
+      newValue: boolean
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'card_is_checkpoint', newValue: boolean
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'linked_skill_id', newValue: string
+  ): void;
+  saveStateProperty(
+      stateName: string, backendName: 'next_content_id_index', newValue: number
+  ): void;
+  saveStateProperty(
+      stateName: string,
+      backendName: StatePropertyNames,
+      newValue: StatePropertyValues
+  ): void {
+    let oldValue = (
+      this.getStatePropertyMemento(stateName, backendName));
+    let newBackendValue = cloneDeep(newValue);
+    let oldBackendValue = cloneDeep(oldValue);
+
+    if (this.BACKEND_CONVERSIONS.hasOwnProperty(backendName)) {
+      newBackendValue = (
+        this.convertToBackendRepresentation(newValue, backendName));
+      oldBackendValue = (
+        this.convertToBackendRepresentation(oldValue, backendName));
+    }
+
+    if (!angular.equals(oldValue, newValue)) {
```

Comment:
Please don't use `angular` global object in migrated code. This object won't be available once angular js is removed.

Lodash can be a replacement here.

Reply:
Done.